File: motif-engine/generate.js
/**
 * Aevia Motif Generator
 * ---------------------
 * Generates small decorative interior motifs for photobook templates
 * using Replicate API.
 *
 * Setup:
 *   node --version   (must be 18+ for built-in fetch)
 *   npm install      (installs replicate package)
 *
 * Usage:
 *   node generate.js
 *   node generate.js --motif tent
 *   node generate.js --motif tent --no-style-image
 */

const fs = require("fs");
const path = require("path");
const https = require("https");

// ── Load .env ────────────────────────────────────────────────────────────────

const envPath = path.join(__dirname, ".env");
if (fs.existsSync(envPath)) {
  for (const line of fs.readFileSync(envPath, "utf8").split("\n")) {
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith("#") && trimmed.includes("=")) {
      const [key, ...rest] = trimmed.split("=");
      process.env[key.trim()] = rest.join("=").trim();
    }
  }
}

const REPLICATE_TOKEN = process.env.REPLICATE_API_TOKEN;
if (!REPLICATE_TOKEN) {
  console.error("ERROR: REPLICATE_API_TOKEN not found in .env file.");
  process.exit(1);
}

// ── Load config ──────────────────────────────────────────────────────────────

const config = JSON.parse(fs.readFileSync(path.join(__dirname, "config.json"), "utf8"));

// ── CLI args ─────────────────────────────────────────────────────────────────

const args = process.argv.slice(2);
const motif = args.includes("--motif")
  ? args[args.indexOf("--motif") + 1]
  : config.active_motif;
const noStyleImage = args.includes("--no-style-image");

const useStyleImage =
  !noStyleImage &&
  config.replicate.style_image_input_field &&
  config.replicate.style_image_reference;

// ── Load prompt ──────────────────────────────────────────────────────────────

const promptPath = path.join(__dirname, "prompts", `${motif}.txt`);
if (!fs.existsSync(promptPath)) {
  console.error(`ERROR: No prompt file found at ${promptPath}`);
  process.exit(1);
}
const promptText = fs.readFileSync(promptPath, "utf8").trim();

console.log(`\nMotif:       ${motif}`);
console.log(`Model:       ${config.replicate.model}`);
console.log(`Style image: ${useStyleImage ? "yes" : "no"}`);
console.log(`Outputs:     ${config.replicate.num_outputs}\n`);

// ── Prepare output folder ────────────────────────────────────────────────────

const today = new Date().toISOString().split("T")[0];
const outputDir = path.join(__dirname, config.output.base_folder, motif, today);
fs.mkdirSync(outputDir, { recursive: true });

const existingFiles = fs.readdirSync(outputDir).filter(f => f.endsWith(".png"));
const existingRuns = existingFiles.map(f => {
  const m = f.match(/_run(\d+)_/);
  return m ? parseInt(m[1]) : 0;
});
const runIndex = existingRuns.length > 0 ? Math.max(...existingRuns) + 1 : 1;
console.log(`Saving to:   ${outputDir}  (run ${runIndex})\n`);

// ── Build model input ────────────────────────────────────────────────────────

const modelInput = {
  prompt: promptText,
  size: config.replicate.size,
  style: config.replicate.style,
  num_outputs: config.replicate.num_outputs,
};

if (useStyleImage) {
  const refFolder = path.join(__dirname, config.references.folder);
  const refImagePath = path.join(refFolder, config.replicate.style_image_reference);

  if (!fs.existsSync(refImagePath)) {
    console.warn(`WARNING: Style image not found at ${refImagePath} — skipping.\n`);
  } else {
    const imageData = fs.readFileSync(refImagePath).toString("base64");
    const ext = path.extname(refImagePath).slice(1).toLowerCase();
    const mime = ext === "jpg" || ext === "jpeg" ? "image/jpeg" : `image/${ext}`;
    modelInput[config.replicate.style_image_input_field] = `data:${mime};base64,${imageData}`;
    console.log(`Style reference: ${path.basename(refImagePath)}\n`);
  }
}

// ── Helper: download file from URL ───────────────────────────────────────────

function downloadFile(url, destPath) {
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(destPath);
    https.get(url, (res) => {
      // Follow redirects
      if (res.statusCode === 301 || res.statusCode === 302) {
        file.close();
        fs.unlinkSync(destPath);
        downloadFile(res.headers.location, destPath).then(resolve).catch(reject);
        return;
      }
      res.pipe(file);
      file.on("finish", () => { file.close(); resolve(); });
    }).on("error", (err) => {
      fs.unlinkSync(destPath);
      reject(err);
    });
  });
}

// ── Call Replicate API ───────────────────────────────────────────────────────

async function run() {
  const Replicate = require("replicate");
  const replicate = new Replicate({ auth: REPLICATE_TOKEN });

  console.log("Calling Replicate... (this may take 20–60 seconds)\n");

  const outputs = await replicate.run(config.replicate.model, { input: modelInput });
  const outputList = Array.isArray(outputs) ? outputs : [outputs];

  console.log(`Got ${outputList.length} outputs.\n`);

  const metadata = {
    motif,
    date: today,
    run: runIndex,
    model: config.replicate.model,
    style_image_used: useStyleImage,
    style_image_reference: useStyleImage ? config.replicate.style_image_reference : null,
    num_outputs: outputList.length,
    files: [],
    prompt: promptText,
  };

  for (let i = 0; i < outputList.length; i++) {
    const output = outputList[i];
    const filename = `${motif}_run${String(runIndex).padStart(2, "0")}_${String(i + 1).padStart(2, "0")}.png`;
    const filepath = path.join(outputDir, filename);

    // output is either a URL string or a FileOutput object
    const url = typeof output === "string" ? output : output.url?.() ?? String(output);
    await downloadFile(url, filepath);

    console.log(`  Saved: ${filename}`);
    metadata.files.push(filename);
  }

  const metaPath = path.join(outputDir, `metadata_run${String(runIndex).padStart(2, "0")}.json`);
  fs.writeFileSync(metaPath, JSON.stringify(metadata, null, 2));

  console.log(`\nDone. Metadata saved to ${path.basename(metaPath)}`);
  console.log(`Open folder: ${outputDir}\n`);
}

run().catch((err) => {
  console.error("\nERROR:", err.message || err);
  process.exit(1);
});

File: motif-engine/generate.py
"""
Aevia Motif Generator
---------------------
Generates small decorative interior motifs for photobook templates
using Replicate API.

Setup:
  pip install replicate
  set REPLICATE_API_TOKEN=your_token_here   (Windows)
  export REPLICATE_API_TOKEN=your_token_here (Mac/Linux)

Usage:
  python generate.py
  python generate.py --motif tent
  python generate.py --motif mountain_peak --no-style-image
"""

import os
import json
import base64
import argparse
import datetime
import urllib.request
from pathlib import Path

# Auto-load .env file from the same folder as this script
# So you just put REPLICATE_API_TOKEN=xxx in motif-engine/.env and it works
def load_env(env_path):
    if not env_path.exists():
        return
    for line in env_path.read_text().splitlines():
        line = line.strip()
        if line and not line.startswith("#") and "=" in line:
            key, _, value = line.partition("=")
            os.environ.setdefault(key.strip(), value.strip())

load_env(Path(__file__).parent / ".env")


# ── Load config ────────────────────────────────────────────────────────────────

SCRIPT_DIR = Path(__file__).parent
config_path = SCRIPT_DIR / "config.json"

with open(config_path) as f:
    config = json.load(f)


# ── CLI args ───────────────────────────────────────────────────────────────────

parser = argparse.ArgumentParser(description="Generate Aevia interior motifs")
parser.add_argument("--motif", default=config["active_motif"],
                    help="Motif name (must match a file in prompts/)")
parser.add_argument("--no-style-image", action="store_true",
                    help="Skip passing the reference image even if configured")
args = parser.parse_args()

motif = args.motif
use_style_image = (
    not args.no_style_image
    and config["replicate"].get("style_image_input_field")
    and config["replicate"].get("style_image_reference")
)


# ── Load prompt ────────────────────────────────────────────────────────────────

prompt_path = SCRIPT_DIR / "prompts" / f"{motif}.txt"
if not prompt_path.exists():
    raise FileNotFoundError(f"No prompt file found at {prompt_path}")

prompt_text = prompt_path.read_text(encoding="utf-8").strip()
print(f"\nMotif: {motif}")
print(f"Model: {config['replicate']['model']}")
print(f"Style image: {'yes' if use_style_image else 'no'}")
print(f"Outputs: {config['replicate']['num_outputs']}\n")


# ── Prepare output folder ──────────────────────────────────────────────────────

date_str = datetime.date.today().strftime("%Y-%m-%d")
output_dir = SCRIPT_DIR / config["output"]["base_folder"] / motif / date_str
output_dir.mkdir(parents=True, exist_ok=True)

# Find the next run index (so reruns don't overwrite)
existing = list(output_dir.glob("*.png"))
run_index = (len(existing) // config["replicate"]["num_outputs"]) + 1
print(f"Saving to: {output_dir}  (run {run_index})\n")


# ── Build model input ──────────────────────────────────────────────────────────

model_input = {
    "prompt": prompt_text,
    "size": config["replicate"]["size"],
    "style": config["replicate"]["style"],
    "num_outputs": config["replicate"]["num_outputs"],
}

# Attach style reference image if configured
if use_style_image:
    ref_folder = SCRIPT_DIR / config["references"]["folder"]
    ref_image_path = ref_folder / config["replicate"]["style_image_reference"]

    if not ref_image_path.exists():
        print(f"WARNING: Style image not found at {ref_image_path} — skipping style image.\n")
    else:
        # Encode image as base64 data URI (works with most Replicate models)
        with open(ref_image_path, "rb") as img_file:
            image_data = base64.b64encode(img_file.read()).decode("utf-8")
        ext = ref_image_path.suffix.lower().replace(".", "")
        mime = "image/jpeg" if ext in ("jpg", "jpeg") else f"image/{ext}"
        field_name = config["replicate"]["style_image_input_field"]
        model_input[field_name] = f"data:{mime};base64,{image_data}"
        print(f"Style reference: {ref_image_path.name}\n")


# ── Call Replicate ─────────────────────────────────────────────────────────────

# Import here so missing package gives a clean error message
try:
    import replicate
except ImportError:
    raise ImportError("Run:  pip install replicate")

print("Calling Replicate... (this may take 20–60 seconds)\n")

outputs = replicate.run(
    config["replicate"]["model"],
    input=model_input
)

# replicate.run() returns a list of output URLs (or file-like objects)
output_urls = list(outputs)
print(f"Got {len(output_urls)} outputs.\n")


# ── Save outputs + metadata ────────────────────────────────────────────────────

metadata = {
    "motif": motif,
    "date": date_str,
    "run": run_index,
    "model": config["replicate"]["model"],
    "style_image_used": use_style_image,
    "style_image_reference": config["replicate"].get("style_image_reference") if use_style_image else None,
    "num_outputs": len(output_urls),
    "files": [],
    "prompt": prompt_text,
}

for i, output in enumerate(output_urls, start=1):
    filename = f"{motif}_run{run_index:02d}_{i:02d}.png"
    filepath = output_dir / filename

    # output can be a URL string or a file-like object depending on replicate version
    if isinstance(output, str):
        urllib.request.urlretrieve(output, filepath)
    else:
        # Newer replicate library returns FileOutput objects
        with open(filepath, "wb") as f:
            f.write(output.read())

    print(f"  Saved: {filename}")
    metadata["files"].append(filename)

# Write metadata JSON alongside the images
meta_path = output_dir / f"metadata_run{run_index:02d}.json"
with open(meta_path, "w") as f:
    json.dump(metadata, f, indent=2)

print(f"\nDone. Metadata saved to {meta_path.name}")
print(f"Open folder: {output_dir}\n")
